server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java

```diff
@@ -393,4 +404,55 @@ private RepositoryData addRandomSnapshotsToRepoData(RepositoryData repoData, boo
         return repoData;
     }
 
+    public void testEnsureUploadListenerIsResolvedWhenAFileSnapshotTaskFails() throws Exception {
+        Settings settings = Settings.builder().put("location", randomAlphaOfLength(10)).build();
+        RepositoryMetadata repositoryMetadata = new RepositoryMetadata(randomAlphaOfLength(10), FsRepository.TYPE, settings);
+        final ClusterService clusterService = BlobStoreTestUtil.mockClusterService(repositoryMetadata);
+        final AtomicBoolean listenerCalled = new AtomicBoolean();
+        final FsRepository repository = new FsRepository(
+            repositoryMetadata,
+            createEnvironment(),
+            xContentRegistry(),
+            clusterService,
+            MockBigArrays.NON_RECYCLING_INSTANCE,
+            new RecoverySettings(Settings.EMPTY, new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS))
+        ) {
+            @Override
+            protected void assertSnapshotOrGenericThread() {
+                // eliminate thread name check as we create repo manually
+            }
+
+            @Override
+            protected void snapshotFile(SnapshotShardContext context, BlobStoreIndexShardSnapshot.FileInfo fileInfo) throws IOException {
+                // Randomly fail some file snapshot tasks
+                if (randomBoolean()) {
+                    throw new IOException("cannot upload file");
+                }
+            }
+        };
+        clusterService.addStateApplier(event -> repository.updateState(event.state()));
+        // Apply state once to initialize repo properly like RepositoriesService would
+        repository.updateState(clusterService.state());
+        repository.start();
+        // Generate some FileInfo, as the files that get uploaded as part of the shard snapshot
+        SnapshotShardContext context = ShardSnapshotTaskRunnerTests.dummyContext();
+        int noOfFiles = randomIntBetween(10, 100);
+        BlockingQueue<BlobStoreIndexShardSnapshot.FileInfo> files = new LinkedBlockingQueue<>(noOfFiles);
+        ActionListener<Collection<Void>> allFilesUploadListener = ActionListener.wrap(() -> listenerCalled.set(true));
+        for (int i = 0; i < noOfFiles; i++) {
+            files.add(ShardSnapshotTaskRunnerTests.dummyFileInfo());
+        }
+        repository.snapshotFiles(context, files, allFilesUploadListener);
+        assertBusy(() -> assertTrue(listenerCalled.get()));
```

Comment:
I kind of dislike this pattern in tests, it causes test to be needlessly slow in aggregate because of the sleeps in the busy polling from `assertTrue`. Maybe just use a `PlainActionFuture` for the listener above and then wait for its completion here?